ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1014,17 +1051,20 @@ void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInf
         BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
         ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
         ByteBuf chunkBuf = blobData.content();
+        this.isChunkCompressed = blobData.isCompressed();
 
         try {
           boolean launchedJob = maybeLaunchCryptoJob(chunkBuf, null, encryptionKey, chunkBlobId);
           if (!launchedJob) {
-            chunkBuf = filterChunkToRange(chunkBuf);
             if (chunkIndex == 0) {
               routerMetrics.getFirstDataChunkLatencyMs.update(System.currentTimeMillis() - submissionTimeMs);
             }
             routerMetrics.getDataChunkLatencyMs.update(System.currentTimeMillis() - initializedTimeMs);
-            chunkIndexToBuf.put(chunkIndex, chunkBuf.retainedDuplicate());
-            numChunksRetrieved.incrementAndGet();
+            ByteBuf decompressedContent = decompressContent(chunkBuf.retainedDuplicate());
+            if (decompressedContent != null) {
+              chunkIndexToBuf.put(chunkIndex, filterChunkToRange(decompressedContent));
```

Comment:
These two lines are common. Can they be pushed down to decompressContent ?

Reply:
No... decompressContent should just decompress content.  Applying range to a chunk is another operation.  If we apply filtering inside decompressContent(), it's considered hidden side-effects which is not desirable.  Also moving that line to the decompressContent() has no performance gain or readability improvement, so I suggest we keep it as is.